src/components/KTreeList/KTreeList.vue

```diff
@@ -0,0 +1,419 @@
+            @change="handleChangeEvent"
+            @selected="handleSelection"
+          >
+            <!-- if TS errors appear on these lines switch to `v-slot:` notation and Save to let linter clear -->
+            <template #[itemIcon]="slotProps">
+              <slot
+                v-bind="slotProps"
+                name="item-icon"
+              />
+            </template>
+            <template #[itemLabel]="slotProps">
+              <slot
+                v-bind="slotProps"
+                name="item-label"
+              />
+            </template>
+          </KTreeList>
+        </div>
+      </template>
+    </draggable>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import { defineProps, defineEmits, computed, ref, watch, onMounted, PropType } from 'vue'
+import draggable from 'vuedraggable'
+import KTreeItem, { TreeListItem } from '@/components/KTreeList/KTreeItem.vue'
+
+export interface ChangeEvent {
+  items: TreeListItem[]
+}
+
+export interface ChildChangeEvent {
+  parent: string,
+  children: TreeListItem[]
+}
+
+const props = defineProps({
+  modelValue: {
+    type: Array as PropType<TreeListItem[]>,
+    default: () => [],
+  },
+  items: {
+    type: Array as PropType<TreeListItem[]>,
+    default: null,
+  },
+  disableDrag: {
+    type: Boolean,
+    default: false,
+  },
+  maxLevel: {
+    type: Number,
+    default: 2,
+  },
+  /**
+     * These are internal props used to track info
+     * of nested children.
+     */
+  level: {
+    type: Number,
+    default: 0,
+  },
+  parentId: {
+    type: String,
+    default: undefined,
+  },
+})
+
+const emit = defineEmits<{
+  (event: 'change', data: ChangeEvent): void,
+  (event: 'child-change', data: ChildChangeEvent): void,
+  (event: 'selected', item: TreeListItem): void
+}>()
+
+const key = ref(0)
+const internalList = ref<TreeListItem[]>([])
+const draggableAttrs = {
+  tag: 'div',
+  // Note: force-fallback: true is required to override cursor when dragging
+  'force-fallback': true,
+  animation: '100',
+  draggable: '.k-tree-item-container',
+  'item-key': 'id',
+  'ghost-class': 'k-tree-item-dragged',
+  'drag-class': 'k-tree-list-grabbing',
+  class: 'child-drop-zone',
+}
+const dragging = ref(false)
+// using these vars to trick the TS compiler and avoid errors about circular refs in the <template>
+const itemIcon = 'item-icon'
+const itemLabel = 'item-label'
+
+// we need this so we can create a watcher for programmatic changes to the modelValue
+const value = computed({
+  get(): TreeListItem[] {
+    return props.modelValue
+  },
+  set(newValue: TreeListItem[]): void {
+    internalList.value = newValue
+  },
+})
+
+const hasChildren = (item: TreeListItem) => {
+  return !!internalList.value.filter(anItem => anItem.id === item.id)?.[0].children?.length
+}
+
+const indentStyle = computed(() => {
+  const level1offset = props.level === 1 ? 6 : 0
+  return {
+    marginLeft: 8 * props.level + level1offset + 'px',
```

Comment:
Couldn't we do this through css pseudo elements? `:fist-child` and `:last-child`?

Reply:
We could do this if we didn't need to calculate the total left margin based off the level, but since we most multiple by the level to get the left margin it doesn't make sense to do some of the calculation in CSS and some of it in JS.

Reply:
I don't understand why we need to calculate the `left-margin` based off of the level? Shouldn't that automatically be taken care when we use the pseudo selector?

Reply:
In your case, applying `margin-left: 8px` to the `.k-tree-item-container` should do the trick and we don't have to calculate margin-left for all children. Also to avoid the margin in the most parent, you can have a `margin-left: -8px` in the `.k-tree-list` class.

Reply:
This solution won't work for nesting multiple levels deep. If we have children 1, 2, and 3 levels deep we are going to have a different margin for each level... maybe I'm not understanding what you are suggesting.

Reply:
It will work for any number of levels. Sorry if it's not making clear. May be refer to the codesandbox link I provided earlier? It has multiple levels and works without any issues.